osu.Game/Online/Chat/MessageFormatter.cs

```diff
@@ -246,6 +247,52 @@ public static LinkDetails GetLinkDetails(string url)
             return new LinkDetails(LinkAction.External, url);
         }
 
+        public static string? GetUrl(LinkDetails link)
+        {
+            switch (link.Action)
+            {
+                case LinkAction.External:
+                    string url = (string)link.Argument;
+
+                    if (url.StartsWith('/'))
+                        url = $"https://{WebsiteRootUrl}{url}";
+
+                    return url;
+
+                case LinkAction.OpenBeatmap:
+                    return $@"https://{WebsiteRootUrl}/beatmaps/{link.Argument}";
+
+                case LinkAction.OpenBeatmapSet:
+                    return $@"https://{WebsiteRootUrl}/beatmapsets/{link.Argument}";
+
+                case LinkAction.OpenUserProfile:
+                    // todo: this is awkward. there should *only* be one type of argument per link action, but alas.
+                    return link.Argument is IUser user
+                        ? $@"https://{WebsiteRootUrl}/users/{user.Username}"
```

Comment:
This should have the online id and fallback to username if it's not populated.

Reply:
I'll much rather put the effort to remove `IUser` from being an argument of `OpenUserProfile` at this point.